### PR TITLE
[Merged by Bors] - feat: lemmas about equiv and logic

### PR DESCRIPTION
**Pull request by fpvandoorn**
* from the Sobolev project (formerly: marginal project)
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Option/Basic.lean [hunk 1]
```diff
@@ -417,4 +417,14 @@ theorem elim_none_some (f : Option α → β) : (fun x ↦ Option.elim x (f none
   funext fun o ↦ by cases o <;> rfl
 #align option.elim_none_some Option.elim_none_some
 
+theorem elim_comp {ι α β} (h : α → β) {f : ι → α} {x : α} {i : Option ι} :
```
Comment (eric-wieser) [inline Mathlib/Data/Option/Basic.lean:420]: Any reason you're not using the ambient `α β γ δ`?

File: Mathlib/Data/Set/Prod.lean [hunk 2]
```diff
@@ -926,3 +934,29 @@ theorem univ_pi_ite (s : Set ι) [DecidablePred (· ∈ s)] (t : ∀ i, Set (α
...
+  convert f.forall_congr_left; rfl
+
+theorem piCongrLeft_preimage_univ_pi (f : ι' ≃ ι) (t : ∀ i, Set (α i)) :
+    f.piCongrLeft α ⁻¹' pi univ t = pi univ fun i => t (f i) := by
+  apply Set.ext; rw [← (f.piCongrLeft α).symm.forall_congr_left]
+  intro x; simp_rw [mem_preimage, apply_symm_apply, mem_univ_pi]
+  exact f.forall_congr_left.symm
+
+theorem piSum_preimage_univ_pi (π : ι ⊕ ι' → Type _) (t : ∀ i, Set (π i)) :
+    piSum π  ⁻¹' pi univ t = pi univ (fun i => t (.inl i)) ×ˢ pi univ fun i => t (.inr i) := by
```
Comment (eric-wieser) [inline Mathlib/Data/Set/Prod.lean:955]:
```suggestion
    piSum π ⁻¹' pi univ t = pi univ (fun i => t (.inl i)) ×ˢ pi univ fun i => t (.inr i) := by
```

File: Mathlib/Data/Set/Prod.lean [hunk 3]
```diff
@@ -926,3 +934,29 @@ theorem univ_pi_ite (s : Set ι) [DecidablePred (· ∈ s)] (t : ∀ i, Set (α
...
 
 end Set
+
+namespace Equiv
+
+open Set
+variable {ι ι' : Type*} {α : ι → Type*}
+
+theorem piCongrLeft_symm_preimage_pi (f : ι' ≃ ι) (s : Set ι) (t : ∀ i, Set (α i)) :
+    ((f.piCongrLeft α).symm ⁻¹' (f ⁻¹' s).pi fun i' => t <| f i') = s.pi t := by
```
Comment (eric-wieser) [inline Mathlib/Data/Set/Prod.lean:944]: Assuming it's still true, this seems like a more useful lemma to me:
```suggestion
    ((f.piCongrLeft α).symm ⁻¹' s.pi fun i' => t <| f i') = (f.symm ⁻¹' s).pi t := by
```

File: Mathlib/Data/Sum/Basic.lean [hunk 4]
```diff
@@ -34,6 +34,10 @@ theorem inl_injective : Function.Injective (inl : α → Sum α β) := fun _ _ 
 theorem inr_injective : Function.Injective (inr : β → Sum α β) := fun _ _ ↦ inr.inj
 #align sum.inr_injective Sum.inr_injective
 
+theorem sum_rec_congr (P : α ⊕ β  → Sort _) (f : ∀ i, P (inl i)) (g : ∀ i, P (inr i))
```
Comment (eric-wieser) [inline Mathlib/Data/Sum/Basic.lean:37]:
```suggestion
theorem sum_rec_congr (P : α ⊕ β → Sort*) (f : ∀ i, P (inl i)) (g : ∀ i, P (inr i))
```

File: Mathlib/Logic/Basic.lean [hunk 5]
```diff
@@ -563,6 +567,9 @@ theorem cast_eq_iff_heq : cast e a = a' ↔ HEq a a' :=
   ⟨heq_of_cast_eq _, fun h ↦ by cases h; rfl⟩
 #align cast_eq_iff_heq cast_eq_iff_heq
 
+theorem Eq.rec_eq_cast {α : Sort _} {P : α → Sort _} {x y : α} (h : x = y) (z : P x) :
+    h ▸ z = cast (congr_arg P h) z := by induction h; rfl
+
```
Comment (eric-wieser) [inline Mathlib/Logic/Basic.lean:572]: I was under the impression that we already had this somewhere...

File: Mathlib/Logic/Equiv/Basic.lean [hunk 6]
```diff
@@ -918,6 +919,21 @@ def arrowProdEquivProdArrow (α β γ : Type*) : (γ → α × β) ≃ (γ → 
 
 open Sum
 
+/-- The type of dependent functions on a sum type `ι ⊕ ι'` is equivalent to the type of pairs of
+functions on `ι` and on `ι'`. This is a dependent version of `Equiv.sumArrowEquivProdArrow`. -/
+@[simps]
+def piSum (π : ι ⊕ ι' → Type _) : ((∀ i, π (inl i)) × ∀ i', π (inr i')) ≃ ∀ i, π i
+    where
```
Comment (eric-wieser) [inline Mathlib/Logic/Equiv/Basic.lean:926]:
```suggestion
functions on `ι` and on `ι'`. This is a dependent version of `Equiv.sumArrowEquivProdArrow`. -/
@[simps]
def sumPiEquivProdPi (π : ι ⊕ ι' → Type _) : (∀ i, π i) ≃ (∀ i, π (inl i)) × (∀ i', π (inr i')) 
    where
```
to match the other name.

File: Mathlib/Logic/Basic.lean (hunk 5 above)
Comment (fpvandoorn) [reply Mathlib/Logic/Basic.lean:572]: `exact?` cannot find anything. `apply?` suggests after two tries:
`((cast_eq_iff_heq.mpr) <| heq_of_eq_rec_left h rfl).symm `

File: Mathlib/Data/Set/Prod.lean (hunk 3 above)
Comment (fpvandoorn) [reply Mathlib/Data/Set/Prod.lean:944]: I agree

File: Mathlib/Logic/Equiv/Basic.lean (hunk 6 above)
Comment (eric-wieser) [reply Mathlib/Logic/Equiv/Basic.lean]: Note I also flipped the direction of the equiv to match, it looks like you haven't done that.

Comment (fpvandoorn) [reply Mathlib/Logic/Equiv/Basic.lean]: Oops, sorry for the mistakes I'm making today.

**Comment by ocfnash** [comment]
Thanks!

bors merge

**Pull request closed without merging**